fdbserver/worker.actor.cpp

```diff
@@ -2257,33 +2258,35 @@ ACTOR Future<MonitorLeaderInfo> monitorLeaderRemotelyOneGeneration(Reference<Clu
 	}
 }
 
-ACTOR Future<Void> monitorLeaderRemotelyInternal(Reference<ClusterConnectionFile> connFile,
-                                                 Reference<AsyncVar<Value>> outSerializedLeaderInfo) {
+ACTOR Future<Void> monitorLeaderWithDelayedCandidacyInternal(Reference<ClusterConnectionFile> connFile,
+                                                             Reference<AsyncVar<Value>> outSerializedLeaderInfo) {
 	state MonitorLeaderInfo info(connFile);
 	loop {
-		MonitorLeaderInfo _info = wait(monitorLeaderRemotelyOneGeneration(connFile, outSerializedLeaderInfo, info));
+		MonitorLeaderInfo _info =
+		    wait(monitorLeaderWithDelayedCandidacyOneGeneration(connFile, outSerializedLeaderInfo, info));
 		info = _info;
 	}
 }
 
 template <class LeaderInterface>
-Future<Void> monitorLeaderRemotely(Reference<ClusterConnectionFile> const& connFile,
-                                   Reference<AsyncVar<Optional<LeaderInterface>>> const& outKnownLeader) {
+Future<Void> monitorLeaderWithDelayedCandidacyImpl(
+    Reference<ClusterConnectionFile> const& connFile,
+    Reference<AsyncVar<Optional<LeaderInterface>>> const& outKnownLeader) {
 	LeaderDeserializer<LeaderInterface> deserializer;
 	auto serializedInfo = makeReference<AsyncVar<Value>>();
-	Future<Void> m = monitorLeaderRemotelyInternal(connFile, serializedInfo);
+	Future<Void> m = monitorLeaderWithDelayedCandidacyInternal(connFile, serializedInfo);
 	return m || deserializer(serializedInfo, outKnownLeader);
 }
 
-ACTOR Future<Void> monitorLeaderRemotelyWithDelayedCandidacy(
+ACTOR Future<Void> monitorLeaderWithDelayedCandidacy(
```

Comment:
I think `WithDelayedCandidacy` only applies to this function. All other `monitorLeaderRemotely*` functions called inside this function doesn't do anything with `DelayedCandidacy`.

I think `DelayedCandidacy` means when this function couldn't get a leader within a timeout, it then tries to become a candidate (line 2316).

And after reading the code, I think `remotely` means it asks the coordinators to run the leader election algorithm and tell this function who is the leader. Comparing to `monitorLeader` function (in this function, it asks all the coordinators who is the nominee, and runs the getLeader() to calculate who is the leader itself), the getLeader() part is run "remotely" in the coordinator.

I'm not sure why these two variants exist, but to me, the difference is that `monitorLeaderRemotely` returns the single true leader (or hang if undecided), and `monitorLeader` returns a list of nominees from each of the coordinators they elect.

Maybe we can give them a better name once having understood why these two variants exist. But I totally agree that these two are very confusing...

Reply:
'MonitorLeaderForProxies' and 'monitorLeaderRemotely' were created as part of the effort to reduce the total number of connections in a cluster. They delegate to a coordinator, I.e. running the leader election and monitoring algorithms on the monitor, each client and SS reduces 'numOfCoord - 1' connections. @sfc-gh-almiller @sfc-gh-etschannen might know more details.

Reply:
Yep, that's pretty much right.  It has to do with trying to reduce connection counts.

monitorLeader directly contacts a quorum of coordinators.  monitorLeaderRemotely is a call that a client makes to one coordinator, and that one coordinator then contacts a quorum of coordinators.  The latter adds an RTT, but is an important optimization for when there's a large number of clients, or when establishing new connections is expensive (TLS).

Reply:
Introduced as part of the aptly named PR #3069: "YOLO at reducing TLS connection count via doing monitorLeader on coordinators"

Reply:
Quite a YOLO. 🚀🌚

Reply:
@halfprice Your understanding is right. I don't like the *Remotely* names because it doesn't clearly describe what the function is doing. Renaming them to *WithDelayedCandidacy* is better because that's a single call path, easier to track. I've also considered *ForSSAndTlog*, but I feel that's too absolute.

Reply:
Thank you @dongxinEric and @alexmiller-apple for the backgroud! @RenxuanW consider adding this info to document the function?

I suspect adding `WithDelayedCandidacy` in function where it doesn't do any `Delay Candidacy` will make it more confusing to future readers. `monitorLeaderWithDelayedCandidacyImpl` maybe fine since it indicates that this is an implementation detail of function `monitorLeaderWithDelayedCandidacy`. But `monitorLeaderWithDelayedCandidacyOneGeneration` probably will cause more confusion. One way is to add `Impl` to all the internal `monitorLeaderWithDelayedCandidacy*` functions, so that it's clear this is an implementation detail of `monitorLeaderWithDelayedCandidacy`, and won't cause mis-use. What do you think?

Reply:
That's fine. Done.